### ISSUE TITLE
"version of Cabal has changed"

### ISSUE THREAD

**Issue by snoyberg** [haskell/cabal#2611]
I know I've reported this issue before on issues that have been closed. There may currently be an open issue on this as well, I'm not certain. This issue has not been resolved, however, and seems to be easy to trigger:

1. Install a version of `cabal-install` with a different Cabal-library major version than what ships with GHC
2. Run `cabal install -j` to do a parallel build
3. Run `cabal build`

This results in the error message:

> cabal: You need to re-run the 'configure' command. The version of Cabal being used has changed (was Cabal-1.18.1.5, now Cabal-1.22.2.0).

```
>ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.8.4
>ghc-pkg list Cabal
WARNING: cache is out of date: C:/Users/Michael/AppData/Local/Programs/minghc-7.8.4-i386/ghc-7.8.4\lib\package.conf.d\package.cache
Use 'ghc-pkg recache' to fix.
C:/Users/Michael/AppData/Local/Programs/minghc-7.8.4-i386/ghc-7.8.4\lib\package.conf.d:
    Cabal-1.18.1.5

C:\Users\Michael\AppData\Roaming\ghc\i386-mingw32-7.8.4\package.conf.d:
    (no packages)
>cabal --version
cabal-install version 1.22.2.0
using version 1.22.2.0 of the Cabal library
```

I performed this test on Windows, but it's occurred to me on Linux many times too.

**Comment by 23Skidoo** [comment]
This is sort of expected behaviour (see #2438, #1938), but since it keeps causing problems, we should add a workaround. I suggest that `setupWrapper` should pick the external setup method even for `build-type: Simple` when the package is already configured and the appropriate version of Cabal is installed.

**Comment by 23Skidoo** [comment]
Alternatively, we can re-run `configure` automatically instead of printing an error message. I believe @ttuegel was looking into making that happen.

**Comment by ttuegel** [comment]
Yes, I have a branch which automatically reconfigures whenever necessary. I just need to do some testing with newer cabal-install vs. older Cabal to be sure they integrate correctly, then I'll make a PR.

**Comment by bitemyapp** [comment]
@ttuegel it would be really lovely to see that go live, thank you! :bear:

**Comment by dcoutts** [comment]
Another fix is here #2633.

**Comment by 23Skidoo** [comment]
Since #2633 has been merged, this can now be closed.
